hardware/VirtualThermostat.cpp

```diff
@@ -0,0 +1,740 @@
+					    SwitchStateAsChanged = true;
+                    }
+					if (  ( abs ( lastPowerModulation - PowerModulation) > 10 )  || ( abs(lastTemp - RoomTemperature ) > 0.2  ) || (SwitchStateAsChanged) )
+					{
+						setOption(Option, "Power"   , PowerModulation);
+						setOption(Option, "RoomTemp", RoomTemperature);
+						setOption(Option, "Switch"  , SwitchValue);     //switch command value
+						m_sql.SetDeviceOptions(ThermostatId, Option );
+
+                        //force display refresh
+                        SendSetPointSensor((uint8_t)(DeviceID>>16) , (DeviceID>>8)&0xFF  , (DeviceID)&0xFF , (float)ThermostatSetPoint, "");
+					}
+				    //if ((Minute % 10 )==0)
+				    //{
+					   // //compute delta room temperature
+					   // float DeltaTemp = RoomTemperature - Map_LastRoomTemp[ThermostatId] ;
+					   // Map_LastRoomTemp[ThermostatId] = RoomTemperature ;
+					   // if ( DeltaTemp == RoomTemperature ) DeltaTemp=0;  //first call
+    				//		TSqlQueryResult res=m_sql.safe_query("SELECT ID FROM DeviceStatus where Name=='his_%s'", ThermostatName ) ;
+    				//		if (res.size() )m_sql.safe_query( "INSERT INTO Temperature (DeviceRowID, Temperature, Humidity ) VALUES (%s , %4.1f, %d )",res[0][0].c_str(),ThermostatSetPoint,PowerModulation	);
+				    //}
+				}
+				else
+					_log.Log(LOG_ERROR,"No switch device associted to Thermostat name =%s", ThermostatName);
+			}
+		}
+		else
+			_log.Log(LOG_ERROR,"No temperature device associted to Thermostat name =%s", ThermostatName);
+	}//nd for
+}
+catch (...)
+{
+	_log.Log(LOG_ERROR,"Exception in ScheduleThermostat");
+}
+}
+
+//return the option value from devive id 
+std::string GetOptionFromDeviceId(std::string &devIdx, const char* option )
+{
+	TOptionMap Option  = m_sql.GetDeviceOptions(devIdx);
+	return (Option[option] );
+} 
+
+//return previous  thermostat target temperature before Time
+int VirtualThermostat::getPrevThermostatProg ( const char * devID , char * CurrentTime , std::string &Time )
+{
+  int TargetTemp=0;
+    TSqlQueryResult result = m_sql.safe_query("SELECT Time,Temperature FROM SetpointTimers where (DeviceRowID==%s) and ( Time < '%s' ) order by time desc limit 1" ,devID, CurrentTime  ) ;
+    if (result.size() ){
+      Time       = result[0][0];
+      TargetTemp =atoi(result[0][1].c_str());
+    }
+return TargetTemp;
+}
+
+//return next  thermostat target temperature after Time
+int VirtualThermostat::getNextThermostatProg ( const char * devID , char * CurrentTime , std::string &Time )
+{
+  int TargetTemp=0;
+    TSqlQueryResult result = m_sql.safe_query("SELECT Time,Temperature FROM SetpointTimers where (DeviceRowID==%s) and ( Time > '%s' ) order by time asc limit 1" ,devID, CurrentTime  ) ;
+    if (result.size() ){
```

Comment:
instead of result.size(), please use

if (!result.empty())

Reply:
ok ! done